File: LeetCode刷题/494.目标和.py
#
# @lc app=leetcode.cn id=494 lang=python3
#
# [494] 目标和
#

# @lc code=start
class Solution:
    def findTargetSumWays(self, nums: list[int], target: int) -> int:
        # 有正有负，也就是说将这些数分成两堆，让他们的和相差为 target；换一个思维就是说有多少种选择方案选择 nums 中的元素，让他们的和为 (sum(nums) + target) // 2；当然从这里可以看出二者的值相加必须为偶数，否则无法构成
        totalSum = sum(nums)
        if (totalSum + target) & 1:
            return 0
        actualTarget = (totalSum+target)//2
        # dp[i][j] 代表的含义是使用 [0, i] 的数构成和 j 的方案数
        # TODO 这个代码还有 bug
        dp = [0] * (1+actualTarget)
        for j in range(0, 1+actualTarget):
            dp[j] = 1 if j == nums[0] else 0
        for i in range(1, len(nums)):
            tmp = [0]*(1+actualTarget)
            for j in range(0, 1+actualTarget):
                if j < nums[i]:
                    tmp[j] = dp[j]
                else:
                    # 如果无法构成，表格中填入的是 0，注意这一点就很简单了
                    tmp[j] = dp[j] + dp[j-nums[i]]
            dp = tmp
        return dp[actualTarget]

# @lc code=end

File: leetcode/cpp-template/leetcode/editor/cn/494.目标和.cpp
/*
 * @lc app=leetcode.cn id=494 lang=cpp
 * @lcpr version=20003
 *
 * [494] 目标和
 */

#include "../common/globalmacros.h"
#include "../common/ListNode.cpp"
#include "../common/TreeNode.cpp"

// @lc code=start




namespace std
{
    template <>
    struct hash<std::pair<int, int>>
    {
        // 哈希函数需要加上限定符 const，否则编译器会报错！
        std::size_t operator()(const std::pair<int, int> &p) const
        {
            return std::hash<int>()(p.first) ^ (std::hash<int>()(p.second) << 1);
        }
    };
}


class Solution
{
public:

    // 解法 1：带 memo 的递归 dp
    // std::pair 中存放的是唯一标识某种情形的 left 和 target
    std::unordered_map<std::pair<int, int>, int> m_memo;

    int findTargetSumWays(std::vector<int> &nums, int target)
    {
        return dp(nums, 0, target);
    }

    int dp(const std::vector<int> &nums, int left, int target)
    {
        if (nums.size() == left) return 0 == target;

        // 先查 memo，查不到递归并写入 memo
        std::pair<int, int> key = {left, target};

        if (0 == m_memo.count(key)) m_memo[key] = dp(nums, 1 + left, target - nums[left]) + dp(nums, 1 + left, target + nums[left]);


        return m_memo[key];
    }
};

// @lc code=end


int main()
{
    Solution solution;
    // your test code here
}


/*
// @lcpr case=start
// [1,1,1,1,1]\n3\n
// @lcpr case=end

// @lcpr case=start
// [1]\n1\n
// @lcpr case=end

 */
